# coding: utf-8

"""
    NCBI Datasets API

    ### NCBI Datasets is a resource that lets you easily gather data from NCBI. The Datasets version 2 API is still in alpha, and we're updating it often to add new functionality, iron out bugs and enhance usability. For some larger downloads, you may want to download a [dehydrated zip archive](https://www.ncbi.nlm.nih.gov/datasets/docs/v2/how-tos/genomes/large-download/), and retrieve the individual data files at a later time. 

    The version of the OpenAPI document: v2
    Generated by OpenAPI Generator (https://openapi-generator.tech)

    Do not edit the class manually.
"""  # noqa: E501


import unittest

from ncbi.datasets.openapi.models.v2reports_gene_descriptor import V2reportsGeneDescriptor

class TestV2reportsGeneDescriptor(unittest.TestCase):
    """V2reportsGeneDescriptor unit test stubs"""

    def setUp(self):
        pass

    def tearDown(self):
        pass

    def make_instance(self, include_optional) -> V2reportsGeneDescriptor:
        """Test V2reportsGeneDescriptor
            include_optional is a boolean, when False only required
            params are included, when True both required and
            optional params are included """
        # uncomment below to create an instance of `V2reportsGeneDescriptor`
        """
        model = V2reportsGeneDescriptor()
        if include_optional:
            return V2reportsGeneDescriptor(
                gene_id = '',
                symbol = '',
                description = '',
                tax_id = '',
                taxname = '',
                common_name = '',
                type = 'UNKNOWN',
                rna_type = 'rna_UNKNOWN',
                orientation = 'none',
                reference_standards = [
                    ncbi.datasets.openapi.models.v2reports_genomic_region.v2reportsGenomicRegion(
                        gene_range = ncbi.datasets.openapi.models.v2reports_seq_range_set.v2reportsSeqRangeSet(
                            accession_version = '', 
                            range = [
                                ncbi.datasets.openapi.models.v2reports_range.v2reportsRange(
                                    begin = '', 
                                    end = '', 
                                    orientation = 'none', 
                                    order = 56, 
                                    ribosomal_slippage = 56, )
                                ], ), 
                        type = 'UNKNOWN', )
                    ],
                genomic_regions = [
                    ncbi.datasets.openapi.models.v2reports_genomic_region.v2reportsGenomicRegion(
                        gene_range = ncbi.datasets.openapi.models.v2reports_seq_range_set.v2reportsSeqRangeSet(
                            accession_version = '', 
                            range = [
                                ncbi.datasets.openapi.models.v2reports_range.v2reportsRange(
                                    begin = '', 
                                    end = '', 
                                    orientation = 'none', 
                                    order = 56, 
                                    ribosomal_slippage = 56, )
                                ], ), 
                        type = 'UNKNOWN', )
                    ],
                chromosomes = [
                    ''
                    ],
                nomenclature_authority = ncbi.datasets.openapi.models.v2reports_nomenclature_authority.v2reportsNomenclatureAuthority(
                    authority = '', 
                    identifier = '', ),
                swiss_prot_accessions = [
                    ''
                    ],
                ensembl_gene_ids = [
                    ''
                    ],
                omim_ids = [
                    ''
                    ],
                synonyms = [
                    ''
                    ],
                replaced_gene_id = '',
                annotations = [
                    ncbi.datasets.openapi.models.v2reports_annotation.v2reportsAnnotation(
                        assembly_accession = '', 
                        assembly_name = '', 
                        annotation_name = '', 
                        annotation_release_date = '', 
                        genomic_locations = [
                            ncbi.datasets.openapi.models.v2reports_genomic_location.v2reportsGenomicLocation(
                                genomic_accession_version = '', 
                                sequence_name = '', 
                                genomic_range = ncbi.datasets.openapi.models.v2reports_range.v2reportsRange(
                                    begin = '', 
                                    end = '', 
                                    orientation = 'none', 
                                    order = 56, 
                                    ribosomal_slippage = 56, ), 
                                exons = [
                                    ncbi.datasets.openapi.models.v2reports_range.v2reportsRange(
                                        begin = '', 
                                        end = '', 
                                        order = 56, 
                                        ribosomal_slippage = 56, )
                                    ], )
                            ], )
                    ],
                transcript_count = 56,
                protein_count = 56,
                transcript_type_counts = [
                    ncbi.datasets.openapi.models.v2reports_transcript_type_count.v2reportsTranscriptTypeCount(
                        type = 'UNKNOWN', 
                        count = 56, )
                    ],
                gene_groups = [
                    ncbi.datasets.openapi.models.v2reports_gene_group.v2reportsGeneGroup(
                        id = '', 
                        method = '', )
                    ],
                summary = [
                    ncbi.datasets.openapi.models.v2reports_gene_summary.v2reportsGeneSummary(
                        source = '', 
                        description = '', 
                        date = '', )
                    ],
                gene_ontology = ncbi.datasets.openapi.models.v2reports_gene_ontology.v2reportsGeneOntology(
                    assigned_by = '', 
                    molecular_functions = [
                        ncbi.datasets.openapi.models.v2reports_process_metadata.v2reportsProcessMetadata(
                            name = '', 
                            go_id = '', 
                            evidence_code = '', 
                            qualifier = '', 
                            reference = ncbi.datasets.openapi.models.v2reports_reference.v2reportsReference(
                                pmids = [
                                    ''
                                    ], ), )
                        ], 
                    biological_processes = [
                        ncbi.datasets.openapi.models.v2reports_process_metadata.v2reportsProcessMetadata(
                            name = '', 
                            go_id = '', 
                            evidence_code = '', 
                            qualifier = '', )
                        ], 
                    cellular_components = [
                        
                        ], ),
                locus_tag = ''
            )
        else:
            return V2reportsGeneDescriptor(
        )
        """

    def testV2reportsGeneDescriptor(self):
        """Test V2reportsGeneDescriptor"""
        # inst_req_only = self.make_instance(include_optional=False)
        # inst_req_and_optional = self.make_instance(include_optional=True)

if __name__ == '__main__':
    unittest.main()
