# coding: utf-8

"""
    NCBI Datasets API

    ### NCBI Datasets is a resource that lets you easily gather data from NCBI. The Datasets version 2 API is still in alpha, and we're updating it often to add new functionality, iron out bugs and enhance usability. For some larger downloads, you may want to download a [dehydrated zip archive](https://www.ncbi.nlm.nih.gov/datasets/docs/v2/how-tos/genomes/large-download/), and retrieve the individual data files at a later time. 

    The version of the OpenAPI document: v2
    Generated by OpenAPI Generator (https://openapi-generator.tech)

    Do not edit the class manually.
"""  # noqa: E501


import unittest

from ncbi.datasets.openapi.models.v2reports_gene_data_report_page import V2reportsGeneDataReportPage

class TestV2reportsGeneDataReportPage(unittest.TestCase):
    """V2reportsGeneDataReportPage unit test stubs"""

    def setUp(self):
        pass

    def tearDown(self):
        pass

    def make_instance(self, include_optional) -> V2reportsGeneDataReportPage:
        """Test V2reportsGeneDataReportPage
            include_optional is a boolean, when False only required
            params are included, when True both required and
            optional params are included """
        # uncomment below to create an instance of `V2reportsGeneDataReportPage`
        """
        model = V2reportsGeneDataReportPage()
        if include_optional:
            return V2reportsGeneDataReportPage(
                reports = [
                    ncbi.datasets.openapi.models.v2reports_gene_report_match.v2reportsGeneReportMatch(
                        gene = ncbi.datasets.openapi.models.v2reports_gene_descriptor.v2reportsGeneDescriptor(
                            gene_id = '', 
                            symbol = '', 
                            description = '', 
                            tax_id = '', 
                            taxname = '', 
                            common_name = '', 
                            type = 'UNKNOWN', 
                            rna_type = 'rna_UNKNOWN', 
                            orientation = 'none', 
                            reference_standards = [
                                ncbi.datasets.openapi.models.v2reports_genomic_region.v2reportsGenomicRegion(
                                    gene_range = ncbi.datasets.openapi.models.v2reports_seq_range_set.v2reportsSeqRangeSet(
                                        accession_version = '', 
                                        range = [
                                            ncbi.datasets.openapi.models.v2reports_range.v2reportsRange(
                                                begin = '', 
                                                end = '', 
                                                order = 56, 
                                                ribosomal_slippage = 56, )
                                            ], ), )
                                ], 
                            genomic_regions = [
                                ncbi.datasets.openapi.models.v2reports_genomic_region.v2reportsGenomicRegion()
                                ], 
                            chromosomes = [
                                ''
                                ], 
                            nomenclature_authority = ncbi.datasets.openapi.models.v2reports_nomenclature_authority.v2reportsNomenclatureAuthority(
                                authority = '', 
                                identifier = '', ), 
                            swiss_prot_accessions = [
                                ''
                                ], 
                            ensembl_gene_ids = [
                                ''
                                ], 
                            omim_ids = [
                                ''
                                ], 
                            synonyms = [
                                ''
                                ], 
                            replaced_gene_id = '', 
                            annotations = [
                                ncbi.datasets.openapi.models.v2reports_annotation.v2reportsAnnotation(
                                    assembly_accession = '', 
                                    assembly_name = '', 
                                    annotation_name = '', 
                                    annotation_release_date = '', 
                                    genomic_locations = [
                                        ncbi.datasets.openapi.models.v2reports_genomic_location.v2reportsGenomicLocation(
                                            genomic_accession_version = '', 
                                            sequence_name = '', 
                                            genomic_range = ncbi.datasets.openapi.models.v2reports_range.v2reportsRange(
                                                begin = '', 
                                                end = '', 
                                                order = 56, 
                                                ribosomal_slippage = 56, ), 
                                            exons = [
                                                
                                                ], )
                                        ], )
                                ], 
                            transcript_count = 56, 
                            protein_count = 56, 
                            transcript_type_counts = [
                                ncbi.datasets.openapi.models.v2reports_transcript_type_count.v2reportsTranscriptTypeCount(
                                    count = 56, )
                                ], 
                            gene_groups = [
                                ncbi.datasets.openapi.models.v2reports_gene_group.v2reportsGeneGroup(
                                    id = '', 
                                    method = '', )
                                ], 
                            summary = [
                                ncbi.datasets.openapi.models.v2reports_gene_summary.v2reportsGeneSummary(
                                    source = '', 
                                    description = '', 
                                    date = '', )
                                ], 
                            gene_ontology = ncbi.datasets.openapi.models.v2reports_gene_ontology.v2reportsGeneOntology(
                                assigned_by = '', 
                                molecular_functions = [
                                    ncbi.datasets.openapi.models.v2reports_process_metadata.v2reportsProcessMetadata(
                                        name = '', 
                                        go_id = '', 
                                        evidence_code = '', 
                                        qualifier = '', 
                                        reference = ncbi.datasets.openapi.models.v2reports_reference.v2reportsReference(
                                            pmids = [
                                                ''
                                                ], ), )
                                    ], 
                                biological_processes = [
                                    ncbi.datasets.openapi.models.v2reports_process_metadata.v2reportsProcessMetadata(
                                        name = '', 
                                        go_id = '', 
                                        evidence_code = '', 
                                        qualifier = '', )
                                    ], 
                                cellular_components = [
                                    
                                    ], ), 
                            locus_tag = '', ), 
                        product = ncbi.datasets.openapi.models.v2reports_product_descriptor.v2reportsProductDescriptor(
                            gene_id = '', 
                            symbol = '', 
                            description = '', 
                            tax_id = '', 
                            taxname = '', 
                            common_name = '', 
                            transcripts = [
                                ncbi.datasets.openapi.models.v2reports_transcript.v2reportsTranscript(
                                    accession_version = '', 
                                    name = '', 
                                    length = 56, 
                                    cds = ncbi.datasets.openapi.models.v2reports_seq_range_set.v2reportsSeqRangeSet(
                                        accession_version = '', ), 
                                    ensembl_transcript = '', 
                                    protein = ncbi.datasets.openapi.models.v2reports_protein.v2reportsProtein(
                                        accession_version = '', 
                                        name = '', 
                                        length = 56, 
                                        isoform_name = '', 
                                        ensembl_protein = '', 
                                        mature_peptides = [
                                            ncbi.datasets.openapi.models.v2reports_mature_peptide.v2reportsMaturePeptide(
                                                accession_version = '', 
                                                name = '', 
                                                length = 56, )
                                            ], ), 
                                    select_category = 'SELECT_UNKNOWN', )
                                ], 
                            transcript_count = 56, 
                            protein_count = 56, ), 
                        query = [
                            ''
                            ], 
                        warnings = [
                            ncbi.datasets.openapi.models.v2reports_warning.v2reportsWarning(
                                gene_warning_code = 'UNKNOWN_GENE_WARNING_CODE', 
                                reason = '', 
                                message = '', 
                                replaced_id = ncbi.datasets.openapi.models.v2reports_warning_replaced_id.v2reportsWarningReplacedId(
                                    requested = '', 
                                    returned = '', ), 
                                unrecognized_identifier = '', )
                            ], 
                        warning = ncbi.datasets.openapi.models.v2reports_warning.v2reportsWarning(
                            reason = '', 
                            message = '', 
                            unrecognized_identifier = '', ), 
                        errors = [
                            ncbi.datasets.openapi.models.v2reports_error.v2reportsError(
                                assembly_error_code = 'UNKNOWN_ASSEMBLY_ERROR_CODE', 
                                gene_error_code = 'UNKNOWN_GENE_ERROR_CODE', 
                                organelle_error_code = 'UNKNOWN_ORGANELLE_ERROR_CODE', 
                                virus_error_code = 'UNKNOWN_VIRUS_ERROR_CODE', 
                                taxonomy_error_code = 'UNKNOWN_TAXONOMY_ERROR_CODE', 
                                reason = '', 
                                message = '', 
                                invalid_identifiers = [
                                    ''
                                    ], )
                            ], )
                    ],
                messages = [
                    ncbi.datasets.openapi.models.v2reports_message.v2reportsMessage(
                        error = ncbi.datasets.openapi.models.v2reports_error.v2reportsError(
                            assembly_error_code = 'UNKNOWN_ASSEMBLY_ERROR_CODE', 
                            gene_error_code = 'UNKNOWN_GENE_ERROR_CODE', 
                            organelle_error_code = 'UNKNOWN_ORGANELLE_ERROR_CODE', 
                            virus_error_code = 'UNKNOWN_VIRUS_ERROR_CODE', 
                            taxonomy_error_code = 'UNKNOWN_TAXONOMY_ERROR_CODE', 
                            reason = '', 
                            message = '', 
                            invalid_identifiers = [
                                ''
                                ], ), 
                        warning = ncbi.datasets.openapi.models.v2reports_warning.v2reportsWarning(
                            gene_warning_code = 'UNKNOWN_GENE_WARNING_CODE', 
                            reason = '', 
                            message = '', 
                            replaced_id = ncbi.datasets.openapi.models.v2reports_warning_replaced_id.v2reportsWarningReplacedId(
                                requested = '', 
                                returned = '', ), 
                            unrecognized_identifier = '', ), )
                    ],
                total_count = 56,
                next_page_token = ''
            )
        else:
            return V2reportsGeneDataReportPage(
        )
        """

    def testV2reportsGeneDataReportPage(self):
        """Test V2reportsGeneDataReportPage"""
        # inst_req_only = self.make_instance(include_optional=False)
        # inst_req_and_optional = self.make_instance(include_optional=True)

if __name__ == '__main__':
    unittest.main()
