# coding: utf-8

"""
    NCBI Datasets API

    ### NCBI Datasets is a resource that lets you easily gather data from NCBI. The Datasets version 2 API is still in alpha, and we're updating it often to add new functionality, iron out bugs and enhance usability. For some larger downloads, you may want to download a [dehydrated zip archive](https://www.ncbi.nlm.nih.gov/datasets/docs/v2/how-tos/genomes/large-download/), and retrieve the individual data files at a later time. 

    The version of the OpenAPI document: v2
    Generated by OpenAPI Generator (https://openapi-generator.tech)

    Do not edit the class manually.
"""  # noqa: E501


import unittest

from ncbi.datasets.openapi.models.v2reports_virus_peptide import V2reportsVirusPeptide

class TestV2reportsVirusPeptide(unittest.TestCase):
    """V2reportsVirusPeptide unit test stubs"""

    def setUp(self):
        pass

    def tearDown(self):
        pass

    def make_instance(self, include_optional) -> V2reportsVirusPeptide:
        """Test V2reportsVirusPeptide
            include_optional is a boolean, when False only required
            params are included, when True both required and
            optional params are included """
        # uncomment below to create an instance of `V2reportsVirusPeptide`
        """
        model = V2reportsVirusPeptide()
        if include_optional:
            return V2reportsVirusPeptide(
                accession = '',
                name = '',
                other_names = [
                    ''
                    ],
                nucleotide = ncbi.datasets.openapi.models.v2reports_seq_range_set_fasta.v2reportsSeqRangeSetFasta(
                    seq_id = '', 
                    accession_version = '', 
                    title = '', 
                    sequence_hash = '', 
                    range = [
                        ncbi.datasets.openapi.models.v2reports_range.v2reportsRange(
                            begin = '', 
                            end = '', 
                            orientation = 'none', 
                            order = 56, 
                            ribosomal_slippage = 56, )
                        ], ),
                protein = ncbi.datasets.openapi.models.v2reports_seq_range_set_fasta.v2reportsSeqRangeSetFasta(
                    seq_id = '', 
                    accession_version = '', 
                    title = '', 
                    sequence_hash = '', 
                    range = [
                        ncbi.datasets.openapi.models.v2reports_range.v2reportsRange(
                            begin = '', 
                            end = '', 
                            orientation = 'none', 
                            order = 56, 
                            ribosomal_slippage = 56, )
                        ], ),
                pdb_ids = [
                    ''
                    ],
                cdd = [
                    ncbi.datasets.openapi.models.v2reports_conserved_domain.v2reportsConservedDomain(
                        accession = '', 
                        name = '', 
                        range = ncbi.datasets.openapi.models.v2reports_range.v2reportsRange(
                            begin = '', 
                            end = '', 
                            orientation = 'none', 
                            order = 56, 
                            ribosomal_slippage = 56, ), )
                    ],
                uni_prot_kb = ncbi.datasets.openapi.models.v2reports_virus_peptide_uni_prot_id.v2reportsVirusPeptideUniProtId(
                    id = '', 
                    name = '', ),
                mature_peptide = [
                    ncbi.datasets.openapi.models.v2reports_virus_peptide.v2reportsVirusPeptide(
                        accession = '', 
                        name = '', 
                        other_names = [
                            ''
                            ], 
                        nucleotide = ncbi.datasets.openapi.models.v2reports_seq_range_set_fasta.v2reportsSeqRangeSetFasta(
                            seq_id = '', 
                            accession_version = '', 
                            title = '', 
                            sequence_hash = '', 
                            range = [
                                ncbi.datasets.openapi.models.v2reports_range.v2reportsRange(
                                    begin = '', 
                                    end = '', 
                                    orientation = 'none', 
                                    order = 56, 
                                    ribosomal_slippage = 56, )
                                ], ), 
                        protein = ncbi.datasets.openapi.models.v2reports_seq_range_set_fasta.v2reportsSeqRangeSetFasta(
                            seq_id = '', 
                            accession_version = '', 
                            title = '', 
                            sequence_hash = '', ), 
                        pdb_ids = [
                            ''
                            ], 
                        cdd = [
                            ncbi.datasets.openapi.models.v2reports_conserved_domain.v2reportsConservedDomain(
                                accession = '', 
                                name = '', )
                            ], 
                        uni_prot_kb = ncbi.datasets.openapi.models.v2reports_virus_peptide_uni_prot_id.v2reportsVirusPeptideUniProtId(
                            id = '', 
                            name = '', ), 
                        protein_completeness = 'UNKNOWN', )
                    ],
                protein_completeness = 'UNKNOWN'
            )
        else:
            return V2reportsVirusPeptide(
        )
        """

    def testV2reportsVirusPeptide(self):
        """Test V2reportsVirusPeptide"""
        # inst_req_only = self.make_instance(include_optional=False)
        # inst_req_and_optional = self.make_instance(include_optional=True)

if __name__ == '__main__':
    unittest.main()
